### PR TITLE
Mysql upsert

### PR DESCRIPTION
**Pull request by mattboyd-aws**
*Issue #, if available:*
https://github.com/awslabs/aws-data-wrangler/issues/608

*Description of changes:*
Adding basic upsert capabilities to `mysql.to_sql()` function. 

`mode` supports 3 new values:
- `upsert_duplicate_key`: Performs an upsert using `ON DUPLICATE KEY` clause. Requires table schema to have defined keys, otherwise duplicate records will be inserted.
- `upsert_replace_into`: Performs upsert using `REPLACE INTO` clause. Less efficient and still requires the table schema to have keys or else duplicate records will be inserted 
- `upsert_distinct`: Inserts new records, including duplicates, then recreates the table and inserts `DISTINCT` records from old table. This is the least efficient approach, but handles scenarios where there are no keys on table.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: awswrangler/mysql.py [hunk 1]
```diff
@@ -284,7 +286,16 @@ def to_sql(
...
         Schema name
     mode : str
-        Append or overwrite.
+        Append, overwrite, upsert_duplicate_key, upsert_replace_into, upsert_distinct.
+            append: Inserts new records into table
+            overwrite: Drops table and recreates
+            upsert_duplicate_key: Performs an upsert using `ON DUPLICATE KEY` clause. Requires table schema to have
+            defined keys, otherwise duplicate records will be inserted.
+            upsert_replace_into: Performs upsert using `REPLACE INTO` clause. Less efficient and still requires the
+            table schema to have keys or else duplicate records will be inserted upsert_distinct: Inserts new records,
```
Comment (jaidisido) [inline awswrangler/mysql.py:295]: Could we bring `upsert_distinct` into a new line for readability?

File: awswrangler/mysql.py [hunk 2]
```diff
@@ -323,6 +334,16 @@ def to_sql(
...
+    if mode.strip().lower() not in [
+        "append",
+        "overwrite",
+        "upsert_replace_into",
+        "upsert_duplicate_key",
+        "upsert_distinct",
+    ]:
+        raise exceptions.InvalidArgumentValue(
+            "mode must be one of append, overwrite, upsert_replace_into, upsert_duplicate_key, upsert_distinct"
+        )
```
Comment (jaidisido) [inline awswrangler/mysql.py:346]: How about we simplify this to:
```
    mode = mode.strip().lower()
    modes = [
        "append",
        "overwrite",
        "upsert_replace_into",
        "upsert_duplicate_key",
        "upsert_distinct",
    ]
    if mode not in modes:
        raise exceptions.InvalidArgumentValue(
            f"mode must be one of {', '.join(modes)}"
        )
```
This way we only defines the list of modes once and likewise only have to perform the `strip().lower()` operation once on mode

File: awswrangler/mysql.py [hunk 3]
```diff
@@ -340,16 +361,33 @@ def to_sql(
...
-                sql: str = f"INSERT INTO `{schema}`.`{table}` {insertion_columns} VALUES {placeholders}"
+                if mode.lower().strip() == "upsert_replace_into":
+                    sql = f"REPLACE INTO `{schema}`.`{table}` {insertion_columns} VALUES {placeholders}"
+                else:
+                    sql = f"INSERT INTO `{schema}`.`{table}` {insertion_columns} VALUES {placeholders}{upsert_str}"
                 _logger.debug("sql: %s", sql)
                 cursor.executemany(sql, (parameters,))
             con.commit()
+            if mode.lower().strip() == "upsert_distinct":
+                temp_table = f"{table}_{''.join(random.choice(string.ascii_lowercase) for i in range(10))}"
```
Comment (jaidisido) [inline awswrangler/mysql.py:384]: `temp_table = f"{table}_{uuid.uuid4().hex}"` is simpler and only requires one library

File: tests/test_mysql.py [hunk 4]
```diff
@@ -200,6 +200,84 @@ def test_insert_with_column_names(mysql_table, mysql_con):
...
+
+    wr.mysql.to_sql(
+        df=df, con=mysql_con, schema="test", table=mysql_table, mode="upsert_distinct", use_column_names=True
+    )
+    wr.mysql.to_sql(
+        df=df, con=mysql_con, schema="test", table=mysql_table, mode="upsert_distinct", use_column_names=True
+    )
+    wr.mysql.to_sql(
+        df=df, con=mysql_con, schema="test", table=mysql_table, mode="upsert_distinct", use_column_names=True
+    )
```
Comment (jaidisido) [inline tests/test_mysql.py:221]: Can we test a few more edge cases for all methods please? For instance what if the primary key values change.
```
    df = pd.DataFrame({"c0": ["foo", "bar"], "c2": [1, 2]})

    wr.mysql.to_sql(
        df=df, con=mysql_con, schema="test", table=mysql_table, mode="upsert_replace_into", use_column_names=True
    )
    df2 = wr.mysql.read_sql_table(con=mysql_con, schema="test", table=mysql_table)
    assert bool(len(df2) == 2)
    wr.mysql.to_sql(
        df=df, con=mysql_con, schema="test", table=mysql_table, mode="upsert_replace_into", use_column_names=True
    )
    df3 = pd.DataFrame({"c0": ["baz", "bar"], "c2": [3, 2]})
    wr.mysql.to_sql(
        df=df3, con=mysql_con, schema="test", table=mysql_table, mode="upsert_replace_into", use_column_names=True
    )
    df4 = wr.mysql.read_sql_table(con=mysql_con, schema="test", table=mysql_table)
    assert bool(len(df4) == 3)
```

**Review by jaidisido — CHANGES_REQUESTED** [review]
Great stuff, thank you for submitting. I left a few comments

File: tests/test_mysql.py (hunk 4 above)
Comment (mattboyd-aws) [reply tests/test_mysql.py:221]: Yes, I will work through more scenarios. Should they be placed into the same test functions with additional assertions or would it be better to have separate functions for each case?

Comment (jaidisido) [reply tests/test_mysql.py:221]: I think additional assertions within the same test functions is fine unless you feel it's getting too big or the cases do not belong to the same logical test. Thanks!

**Review by senorkrabs — APPROVED** [review]

**Review by senorkrabs — APPROVED** [review]

**Comment by senorkrabs** [comment]
The latest commit should have the requested changes. Please let me know if I missed anything.

File: awswrangler/mysql.py [hunk 5]
```diff
@@ -340,16 +362,33 @@ def to_sql(
                 df.reset_index(level=df.index.names, inplace=True)
             column_placeholders: str = ", ".join(["%s"] * len(df.columns))
             insertion_columns = ""
+            upsert_columns = ""
+            upsert_str = ""
             if use_column_names:
                 insertion_columns = f"({', '.join(df.columns)})"
+            if mode.lower().strip() == "upsert_duplicate_key":
```
Comment (jaidisido) [inline awswrangler/mysql.py:369]: Since the `lower().strip()` operation was already performed once, there is no need to run it here again

File: awswrangler/mysql.py [hunk 6]
```diff
@@ -340,16 +362,33 @@ def to_sql(
...
+                upsert_columns = ", ".join(df.columns.map(lambda column: f"`{column}`=VALUES(`{column}`)"))
+                upsert_str = f" ON DUPLICATE KEY UPDATE {upsert_columns}"
             placeholder_parameter_pair_generator = _db_utils.generate_placeholder_parameter_pairs(
                 df=df, column_placeholders=column_placeholders, chunksize=chunksize
             )
+            sql: str
             for placeholders, parameters in placeholder_parameter_pair_generator:
-                sql: str = f"INSERT INTO `{schema}`.`{table}` {insertion_columns} VALUES {placeholders}"
+                if mode.lower().strip() == "upsert_replace_into":
```
Comment (jaidisido) [inline awswrangler/mysql.py:377]: here

File: awswrangler/mysql.py [hunk 7]
```diff
@@ -340,16 +362,33 @@ def to_sql(
...
+                    sql = f"REPLACE INTO `{schema}`.`{table}` {insertion_columns} VALUES {placeholders}"
+                else:
+                    sql = f"INSERT INTO `{schema}`.`{table}` {insertion_columns} VALUES {placeholders}{upsert_str}"
                 _logger.debug("sql: %s", sql)
                 cursor.executemany(sql, (parameters,))
             con.commit()
+            if mode.lower().strip() == "upsert_distinct":
```
Comment (jaidisido) [inline awswrangler/mysql.py:384]: and here

**Review by jaidisido — CHANGES_REQUESTED** [review]
Minor change requested otherwise looks great, thanks

**Review by jaidisido — APPROVED** [review]
